pkg/storage/ingest/reader.go

```diff
@@ -184,12 +184,13 @@ func (r *PartitionReader) processNextFetches(ctx context.Context, delayObserver
 	r.logFetchErrors(fetches)
 	fetches = filterOutErrFetches(fetches)
 
-	// TODO consumeFetches() may get interrupted in the middle because of ctx canceled due to PartitionReader stopped.
-	// 		We should improve it, but we shouldn't just pass a context.Background() because if consumption is stuck
-	// 		then PartitionReader will never stop.
-	r.consumeFetches(ctx, fetches)
-	r.enqueueCommit(fetches)
-	r.notifyLastConsumedOffset(fetches)
+	err := r.consumeFetches(ctx, fetches)
+	if err != nil {
+		level.Error(r.logger).Log("msg", "error while consuming fetches", "err", err, "num_records", fetches.NumRecords())
```

Comment:
I'm not sure we log as error when the context is canceled (which is the only error that could be returned). I would rather propagate the error back to the caller, up until `run()` and then in `run()` we check whether it's context canceled or a different error. If it's context canceled, we interrupt `run()` returning `nil`, while if it's an other we return the `err` (this will make the entire ingester to fail, which is what we want in case of a non recoverable). I know we currently can only return `context.Canceled` but I think will make code more future proof.